File: CPP/matrix/by_def.cpp
#include <vector>

using namespace std;

bool valid(vector<vector<int>> A, vector<vector<int>> B)
{
	if(A.size() > 0 && B.size() > 0 &&
			A[0].size() > 0 && B[0].size() > 0 &&
			A[0].size() == B.size())
	{
		return true;
	}
	else
	{
		return false;
	}
}

vector<vector<int>> multiply(vector<vector<int>> A, vector<vector<int>> B)
{
	if(!valid(A,B)) return {};

	size_t m = A.size();
	size_t n = A[0].size();
	size_t p = B[0].size();

	vector<vector<int>> C(m, vector<int> (p));

	for(int i = 0; i < m; i++)
	{
		for(int k = 0; k < p; k++)
		{
			for(int j = 0; j < n; j++)
			{
				C[i][k] += A[i][j]*B[j][k];
			}
		}
	}
	return C;
}

File: Rust/matrix/by_def.rs
use std::vec;

fn valid(A: Vec<Vec<i32>>, B: Vec<Vec<i32>>) -> bool
{
    if A.len() > 0 && B.len() > 0 && A[0].len() > 0 && B[0].len() > 0 && A[0].len() == B.len()
    {
        return true;
    }
    else
    {
        return false;
    }
}


fn multiply(A: Vec<Vec<i32>>, B: Vec<Vec<i32>>) -> Vec<Vec<i32>>
{
    if !(valid(A,B))
    {
        return vec![];
    }

    let m = A.len();
    let n = A[0].len();
    let p = B[0].len();

    let mut C = vec![vec![0; m]; p];

    for i in 0..m
    {
        for k in 0..p
        {
            for j in 0..n
            {
                C[i][k] += A[i][j]*B[j][k];
            }
        }
    }
    return C;
}
